src/libraries/Common/src/Interop/SunOS/procfs/Interop.ProcFs.GetProcessInfoById.cs

```diff
@@ -0,0 +1,96 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Diagnostics;
+using System.Diagnostics.CodeAnalysis;
+using System.Runtime.InteropServices;
+using System.IO;
+
+internal static partial class Interop
+{
+    internal static partial class @procfs
+    {
+
+        /// <summary>
+        /// Attempts to get status info for the specified process ID.
+        /// </summary>
+        /// <param name="pid">PID of the process to read status info for.</param>
+        /// <param name="result">The pointer to ProcessInfo instance.</param>
+        /// <returns>
+        /// true if the process status was read; otherwise, false.
+        /// </returns>
+
+        // ProcessManager.SunOS.cs calls this
+        // "unsafe" due to use of fixed-size buffers
+
+        internal static unsafe bool GetProcessInfoById(int pid, out ProcessInfo result)
+        {
+            result = default;
+            bool ret = false;
+            string fileName = "?";
+            IntPtr ptr = 0;
+
+            try
+            {
+                fileName = GetInfoFilePathForProcess(pid);
+                int size = Marshal.SizeOf<psinfo>();
+                ptr = Marshal.AllocHGlobal(size);
+
+                BinaryReader br = new BinaryReader(File.OpenRead(fileName));
+                byte[] buf = br.ReadBytes(size);
+                Marshal.Copy(buf, 0, ptr, size);
+
+                procfs.psinfo pr = Marshal.PtrToStructure<psinfo>(ptr);
+
+                GetProcessInfoFromInternal(ref result, ref pr);
+                // Had trouble with CS1666 if I move this into GetProcessInfoFromInternal
+                result.Args = Marshal.PtrToStringUTF8((IntPtr)pr.pr_psargs);
+
+                // We get LWP[1] for "free"
+                GetThreadInfoFromInternal(ref result.Lwp1, ref pr.pr_lwp);
+
+                ret = true;
+            }
+            catch (Exception e)
+            {
+                Debug.Fail($"Failed to read \"{fileName}\": {e}");
+            }
+            finally
+            {
+                Marshal.FreeHGlobal(ptr);
+            }
+
+            return ret;
+        }
```

Comment:
This implementation is allocation unnecessary stuff and we shouldn't be reading unused stuff for environment.sunos:

 ```suggestion
internal static unsafe bool TryGetProcessInfoById(int pid, out ProcessInfo result)
{
    result = default;
    bool ret = false;

    int size = sizeof(psinfo);
    Debug.Assert(size <= 1024, "psinfo struct size exceeds 1024 bytes.");

    byte* buffer = stackalloc byte[size];

    try
    {
        string fileName = GetInfoFilePathForProcess(pid);

        using (FileStream fs = new FileStream(fileName, FileMode.Open, FileAccess.Read))
        {
            int bytesRead = fs.Read(new Span<byte>(buffer, size));
            if (bytesRead != size)
            {
                throw new IOException($"Failed to read the entire psinfo structure from {fileName}");
            }
        }

        psinfo* pr = (psinfo*)buffer;

        result.Pid = pr->pr_pid;
        result.ParentPid = pr->pr_ppid;
        result.SessionId = pr->pr_sid;
        result.VirtualSize = (nuint)pr->pr_size * 1024;
        result.ResidentSetSize = (nuint)pr->pr_rssize * 1024;
        result.StartTime.TvSec = pr->pr_start.tv_sec;
        result.StartTime.TvNsec = pr->pr_start.tv_nsec;
        result.CpuTotalTime.TvSec = pr->pr_time.tv_sec;
        result.CpuTotalTime.TvNsec = pr->pr_time.tv_nsec;
        result.Args = Marshal.PtrToStringUTF8((IntPtr)pr->pr_psargs);

        GetThreadInfoFromInternal(ref result.Lwp1, ref pr->pr_lwp);

        ret = true;
    }
    catch (Exception e)
    {
        Debug.Fail($"Failed to read process info: {e}");
    }

    return ret;
}

internal static unsafe long TryGetResidentSetSizeById(int pid)
{
    int size = sizeof(psinfo);
    Debug.Assert(size <= 1024, "psinfo struct size exceeds 1024 bytes.");

    byte* buffer = stackalloc byte[size];

    try
    {
        string fileName = GetInfoFilePathForProcess(pid);

        using (FileStream fs = new FileStream(fileName, FileMode.Open, FileAccess.Read))
        {
            int bytesRead = fs.Read(new Span<byte>(buffer, size));
            if (bytesRead != size)
            {
                throw new IOException($"Failed to read the entire psinfo structure from {fileName}");
            }
        }

        psinfo* pr = (psinfo*)buffer;

        // The Resident Set Size (RSS) in bytes
        return (long)pr->pr_rssize * 1024;
    }
    catch (Exception e)
    {
        Debug.Fail($"Failed to read process info: {e}");
        return 0;
    }
}
```

Reply:
What's the reason for wanting `Try...` at the front of the names?
Are the allocations for the ProcessInfo object worth worrying about here?
I was also wondering if the environment.sunos call could do something

Reply:
> This implementation is allocating unnecessary stuff and we shouldn't be reading unused stuff for environment.sunos. Also, it is missing `Try` prefix.

Huh.  So we don't need all the BinaryReader and Marshal.method stuff?
I'll try your version and see how it goes...

Reply:
Can you please put those suggestions on a branch that I can fetch?  Thanks.

Reply:
You can apply the suggestion from GitHub UI.
and `Try`-prefix is a common convention in C# for the method returning `bool`, handles an exception and has an `out` argument.

Reply:
I want to try the changes locally, and once I'm happy, then update the PR.

Reply:
OK, downloaded https://github.com/am11/runtime/commit/b018235f34268b5f69ec259ae73adc30336f27dd.patch
Thanks.